File: rust/tests/cli_options.rs
//! Integration tests for all CLI options.
//!
//! Verifies that the Rust CLI binary supports the same set of CLI options
//! as the JavaScript implementation, ensuring full feature parity.
//! Each option is tested via the compiled binary using assert_cmd.

use assert_cmd::Command;
use predicates::prelude::*;
use std::io::Write;

/// Helper to create a Command for the agent binary.
fn agent_cmd() -> assert_cmd::Command {
    Command::cargo_bin("agent").unwrap()
}

// ── Model option ──────────────────────────────────────────────────────

#[test]
fn model_option_default() {
    // Default model should be opencode/nemotron-3-super-free (matching JS).
    agent_cmd()
        .args(["--dry-run", "--verbose", "-p", "hello"])
        .assert()
        .success()
        .stdout(predicate::str::contains(
            "Model: opencode/nemotron-3-super-free",
        ));
}

#[test]
fn model_option_custom() {
    agent_cmd()
        .args([
            "--dry-run",
            "--verbose",
            "--model",
            "opencode/gpt-5",
            "-p",
            "hello",
        ])
        .assert()
        .success()
        .stdout(predicate::str::contains("Model: opencode/gpt-5"));
}

// ── JSON standard option ─────────────────────────────────────────────

#[test]
fn json_standard_default() {
    agent_cmd()
        .args(["--dry-run", "--verbose", "-p", "hello"])
        .assert()
        .success()
        .stdout(predicate::str::contains("JSON standard: opencode"));
}

#[test]
fn json_standard_claude() {
    agent_cmd()
        .args([
            "--dry-run",
            "--verbose",
            "--json-standard",
            "claude",
            "-p",
            "hello",
        ])
        .assert()
        .success()
        .stdout(predicate::str::contains("JSON standard: claude"));
}

#[test]
fn json_standard_rejects_invalid() {
    agent_cmd()
        .args(["--dry-run", "--json-standard", "xml", "-p", "hello"])
        .assert()
        .failure()
        .stderr(predicate::str::contains("invalid value"));
}

// ── System message options ───────────────────────────────────────────

#[test]
fn system_message_option() {
    agent_cmd()
        .args([
            "--dry-run",
            "--verbose",
            "--system-message",
            "You are a test bot",
            "-p",
            "hello",
        ])
        .assert()
        .success()
        .stdout(predicate::str::contains(
            "System message: You are a test bot",
        ));
}

#[test]
fn system_message_file_option() {
    // Create a temp file with system message content
    let mut tmpfile = tempfile::NamedTempFile::new().unwrap();
    write!(tmpfile, "System from file").unwrap();
    let path = tmpfile.path().to_str().unwrap().to_string();

    agent_cmd()
        .args([
            "--dry-run",
            "--verbose",
            "--system-message-file",
            &path,
            "-p",
            "hello",
        ])
        .assert()
        .success()
        .stdout(predicate::str::contains("System message: System from file"));
}

#[test]
fn append_system_message_option() {
    agent_cmd()
        .args([
            "--dry-run",
            "--verbose",
            "--append-system-message",
            "Extra instructions",
            "-p",
            "hello",
        ])
        .assert()
        .success()
        .stdout(predicate::str::contains(
            "Append system message: Extra instructions",
        ));
}

#[test]
fn append_system_message_file_option() {
    let mut tmpfile = tempfile::NamedTempFile::new().unwrap();
    write!(tmpfile, "Appended from file").unwrap();
    let path = tmpfile.path().to_str().unwrap().to_string();

    agent_cmd()
        .args([
            "--dry-run",
            "--verbose",
            "--append-system-message-file",
            &path,
            "-p",
            "hello",
        ])
        .assert()
        .success()
        .stdout(predicate::str::contains(
            "Append system message: Appended from file",
        ));
}

#[test]
fn system_message_file_not_found() {
    agent_cmd()
        .args([
            "--dry-run",
            "--system-message-file",
            "/tmp/nonexistent_file_12345.txt",
            "-p",
            "hello",
        ])
        .assert()
        .failure();
}

// ── Server mode option ───────────────────────────────────────────────

#[test]
fn server_mode_default_true() {
    agent_cmd()
        .args(["--dry-run", "--verbose", "-p", "hello"])
        .assert()
        .success()
        .stdout(predicate::str::contains("Server mode: true"));
}

#[test]
fn server_mode_disabled() {
    agent_cmd()
        .args(["--dry-run", "--verbose", "--no-server", "-p", "hello"])
        .assert()
        .success()
        .stdout(predicate::str::contains("Server mode: false"));
}

// ── Verbose option ───────────────────────────────────────────────────

#[test]
fn verbose_shows_config() {
    // When verbose is on, many config values should be displayed.
    agent_cmd()
        .args(["--dry-run", "--verbose", "-p", "hello"])
        .assert()
        .success()
        .stdout(predicate::str::contains("Model:"))
        .stdout(predicate::str::contains("JSON standard:"))
        .stdout(predicate::str::contains("Compaction model:"))
        .stdout(predicate::str::contains("Compaction models:"))
        .stdout(predicate::str::contains("Compaction safety margin:"))
        .stdout(predicate::str::contains("Interactive:"))
        .stdout(predicate::str::contains("Generate title:"))
        .stdout(predicate::str::contains("Summarize session:"))
        .stdout(predicate::str::contains("Retry on rate limits:"));
}

#[test]
fn verbose_off_hides_config() {
    // Without verbose, config lines should NOT appear.
    agent_cmd()
        .args(["--dry-run", "-p", "hello"])
        .assert()
        .success()
        .stdout(predicate::str::contains("Model:").not())
        .stdout(predicate::str::contains("JSON standard:").not());
}

// ── Dry run option ───────────────────────────────────────────────────

#[test]
fn dry_run_echoes_message() {
    agent_cmd()
        .args(["--dry-run", "-p", "test message"])
        .assert()
        .success()
        .stdout(predicate::str::contains(
            "[DRY RUN] Received message: test message",
        ));
}

// ── Use existing Claude OAuth option ─────────────────────────────────

#[test]
fn use_existing_claude_oauth_accepted() {
    agent_cmd()
        .args(["--dry-run", "--use-existing-claude-oauth", "-p", "hello"])
        .assert()
        .success();
}

// ── Prompt option ────────────────────────────────────────────────────

#[test]
fn prompt_short_flag() {
    agent_cmd()
        .args(["--dry-run", "-p", "short flag test"])
        .assert()
        .success()
        .stdout(predicate::str::contains("short flag test"));
}

#[test]
fn prompt_long_flag() {
    agent_cmd()
        .args(["--dry-run", "--prompt", "long flag test"])
        .assert()
        .success()
        .stdout(predicate::str::contains("long flag test"));
}

// ── Disable stdin option ─────────────────────────────────────────────

#[test]
fn disable_stdin_with_prompt_succeeds() {
    agent_cmd()
        .args(["--dry-run", "--disable-stdin", "-p", "hello"])
        .assert()
        .success();
}

#[test]
fn disable_stdin_without_prompt_fails() {
    agent_cmd()
        .args(["--dry-run", "--disable-stdin"])
        .assert()
        .failure();
}

// ── Stdin stream timeout option ──────────────────────────────────────

#[test]
fn stdin_stream_timeout_accepted() {
    agent_cmd()
        .args(["--dry-run", "--stdin-stream-timeout", "5000", "-p", "hello"])
        .assert()
        .success();
}

// ── Auto merge queued messages option ────────────────────────────────

#[test]
fn auto_merge_queued_messages_default() {
    agent_cmd()
        .args(["--dry-run", "-p", "hello"])
        .assert()
        .success();
}

#[test]
fn no_auto_merge_queued_messages() {
    agent_cmd()
        .args([
            "--dry-run",
            "--no-auto-merge-queued-messages",
            "-p",
            "hello",
        ])
        .assert()
        .success();
}

// ── Interactive option ───────────────────────────────────────────────

#[test]
fn interactive_default_true() {
    agent_cmd()
        .args(["--dry-run", "--verbose", "-p", "hello"])
        .assert()
        .success()
        .stdout(predicate::str::contains("Interactive: true"));
}

#[test]
fn no_interactive() {
    agent_cmd()
        .args(["--dry-run", "--verbose", "--no-interactive", "-p", "hello"])
        .assert()
        .success()
        .stdout(predicate::str::contains("Interactive: false"));
}

// ── Always accept stdin option ───────────────────────────────────────

#[test]
fn no_always_accept_stdin() {
    agent_cmd()
        .args(["--dry-run", "--no-always-accept-stdin", "-p", "hello"])
        .assert()
        .success();
}

// ── Compact JSON option ──────────────────────────────────────────────

#[test]
fn compact_json_single_line() {
    let output = agent_cmd()
        .args(["--dry-run", "--compact-json", "-p", "hello"])
        .output()
        .unwrap();
    let stdout = String::from_utf8(output.stdout).unwrap();
    // In compact mode, each JSON event should be on a single line
    for line in stdout.lines() {
        if line.trim().is_empty() {
            continue;
        }
        // Each non-empty line should be valid JSON
        assert!(
            serde_json::from_str::<serde_json::Value>(line).is_ok(),
            "Expected valid JSON on single line, got: {}",
            line
        );
    }
}

// ── Resume option ────────────────────────────────────────────────────

#[test]
fn resume_option_accepted() {
    agent_cmd()
        .args(["--dry-run", "--resume", "ses_abc123", "-p", "hello"])
        .assert()
        .success();
}

#[test]
fn resume_short_flag() {
    agent_cmd()
        .args(["--dry-run", "-r", "ses_abc123", "-p", "hello"])
        .assert()
        .success();
}

// ── Continue option ──────────────────────────────────────────────────

#[test]
fn continue_option_accepted() {
    agent_cmd()
        .args(["--dry-run", "--continue", "-p", "hello"])
        .assert()
        .success();
}

#[test]
fn continue_short_flag() {
    agent_cmd()
        .args(["--dry-run", "-c", "-p", "hello"])
        .assert()
        .success();
}

// ── No fork option ───────────────────────────────────────────────────

#[test]
fn no_fork_option_accepted() {
    agent_cmd()
        .args([
            "--dry-run",
            "--no-fork",
            "--resume",
            "ses_abc",
            "-p",
            "hello",
        ])
        .assert()
        .success();
}

// ── Generate title option ────────────────────────────────────────────

#[test]
fn generate_title_option() {
    agent_cmd()
        .args(["--dry-run", "--verbose", "--generate-title", "-p", "hello"])
        .assert()
        .success()
        .stdout(predicate::str::contains("Generate title: true"));
}

// ── Retry timeout option ─────────────────────────────────────────────

#[test]
fn retry_timeout_option() {
    agent_cmd()
        .args([
            "--dry-run",
            "--verbose",
            "--retry-timeout",
            "3600",
            "-p",
            "hello",
        ])
        .assert()
        .success()
        .stdout(predicate::str::contains("Retry timeout: 3600s"));
}

// ── Retry on rate limits option ──────────────────────────────────────

#[test]
fn retry_on_rate_limits_default_true() {
    agent_cmd()
        .args(["--dry-run", "--verbose", "-p", "hello"])
        .assert()
        .success()
        .stdout(predicate::str::contains("Retry on rate limits: true"));
}

#[test]
fn no_retry_on_rate_limits() {
    agent_cmd()
        .args([
            "--dry-run",
            "--verbose",
            "--no-retry-on-rate-limits",
            "-p",
            "hello",
        ])
        .assert()
        .success()
        .stdout(predicate::str::contains("Retry on rate limits: false"));
}

// ── Output response model option ─────────────────────────────────────

#[test]
fn output_response_model_accepted() {
    agent_cmd()
        .args(["--dry-run", "--no-output-response-model", "-p", "hello"])
        .assert()
        .success();
}

// ── Summarize session option ─────────────────────────────────────────

#[test]
fn summarize_session_default() {
    agent_cmd()
        .args(["--dry-run", "--verbose", "-p", "hello"])
        .assert()
        .success()
        .stdout(predicate::str::contains("Summarize session: true"));
}

#[test]
fn no_summarize_session() {
    agent_cmd()
        .args([
            "--dry-run",
            "--verbose",
            "--no-summarize-session",
            "-p",
            "hello",
        ])
        .assert()
        .success()
        .stdout(predicate::str::contains("Summarize session: false"));
}

// ── Compaction model option ──────────────────────────────────────────

#[test]
fn compaction_model_default() {
    agent_cmd()
        .args(["--dry-run", "--verbose", "-p", "hello"])
        .assert()
        .success()
        .stdout(predicate::str::contains(
            "Compaction model: opencode/gpt-5-nano",
        ));
}

#[test]
fn compaction_model_custom() {
    agent_cmd()
        .args([
            "--dry-run",
            "--verbose",
            "--compaction-model",
            "same",
            "-p",
            "hello",
        ])
        .assert()
        .success()
        .stdout(predicate::str::contains("Compaction model: same"));
}

// ── Compaction models option ─────────────────────────────────────────

#[test]
fn compaction_models_default() {
    agent_cmd()
        .args(["--dry-run", "--verbose", "-p", "hello"])
        .assert()
        .success()
        .stdout(predicate::str::contains(
            "Compaction models: (big-pickle minimax-m2.5-free nemotron-3-super-free gpt-5-nano same)",
        ));
}

#[test]
fn compaction_models_custom() {
    agent_cmd()
        .args([
            "--dry-run",
            "--verbose",
            "--compaction-models",
            "(model1 same)",
            "-p",
            "hello",
        ])
        .assert()
        .success()
        .stdout(predicate::str::contains("Compaction models: (model1 same)"));
}

// ── Compaction safety margin option ──────────────────────────────────

#[test]
fn compaction_safety_margin_default() {
    agent_cmd()
        .args(["--dry-run", "--verbose", "-p", "hello"])
        .assert()
        .success()
        .stdout(predicate::str::contains("Compaction safety margin: 25%"));
}

#[test]
fn compaction_safety_margin_custom() {
    agent_cmd()
        .args([
            "--dry-run",
            "--verbose",
            "--compaction-safety-margin",
            "25",
            "-p",
            "hello",
        ])
        .assert()
        .success()
        .stdout(predicate::str::contains("Compaction safety margin: 25%"));
}

// ── All options combined ─────────────────────────────────────────────

#[test]
fn all_options_accepted_together() {
    // Verify that the binary accepts all options simultaneously without conflict.
    agent_cmd()
        .args([
            "--model",
            "opencode/gpt-5",
            "--json-standard",
            "claude",
            "--system-message",
            "Be helpful",
            "--verbose",
            "--dry-run",
            "--use-existing-claude-oauth",
            "--compact-json",
            "--no-interactive",
            "--no-always-accept-stdin",
            "--no-auto-merge-queued-messages",
            "--generate-title",
            "--no-retry-on-rate-limits",
            "--retry-timeout",
            "60",
            "--no-output-response-model",
            "--no-summarize-session",
            "--compaction-model",
            "same",
            "--compaction-models",
            "(same)",
            "--compaction-safety-margin",
            "20",
            "--temperature",
            "0.5",
            "--no-server",
            "--no-fork",
            "--resume",
            "ses_abc",
            "--stdin-stream-timeout",
            "1000",
            "--disable-stdin",
            "-p",
            "test",
        ])
        .assert()
        .success();
}

File: js/tests/cli_options.ts
import { describe, expect, test } from 'bun:test';
import yargs from 'yargs';
import { buildRunOptions } from '../src/cli/run-options.js';
import {
  DEFAULT_COMPACTION_MODEL_ENV,
  DEFAULT_COMPACTION_MODELS_ENV,
  DEFAULT_COMPACTION_SAFETY_MARGIN_PERCENT_ENV,
  DEFAULT_MODEL_ENV,
} from '../src/config/defaults';

const defaultEnvKeys = [
  DEFAULT_MODEL_ENV,
  DEFAULT_COMPACTION_MODEL_ENV,
  DEFAULT_COMPACTION_MODELS_ENV,
  DEFAULT_COMPACTION_SAFETY_MARGIN_PERCENT_ENV,
];

function testEnv(
  overrides: Record<string, string> = {}
): Record<string, string> {
  const env: Record<string, string> = {};
  for (const [key, value] of Object.entries(process.env)) {
    if (value !== undefined) env[key] = value;
  }
  for (const key of defaultEnvKeys) {
    delete env[key];
  }
  return { ...env, ...overrides };
}

async function parseRunOptions(
  args: string[],
  envOverrides: Record<string, string> = {}
): Promise<Record<string, any>> {
  const parser = buildRunOptions(yargs(args), {
    env: testEnv(envOverrides),
  })
    .scriptName('agent')
    .exitProcess(false)
    .help(false)
    .fail((message, error) => {
      throw error ?? new Error(message);
    });

  return (await parser.parse()) as Record<string, any>;
}

describe('cli_options', () => {
  test('model_option_default', async () => {
    const argv = await parseRunOptions([]);
    expect(argv.model).toBe('opencode/minimax-m2.5-free');
  });

  test('model_option_custom', async () => {
    const argv = await parseRunOptions(['--model', 'opencode/gpt-5']);
    expect(argv.model).toBe('opencode/gpt-5');
  });

  test('model_option_env_default', async () => {
    const argv = await parseRunOptions([], {
      [DEFAULT_MODEL_ENV]: 'opencode/env-default-free',
    });
    expect(argv.model).toBe('opencode/env-default-free');
  });

  test('model_option_cli_overrides_env_default', async () => {
    const argv = await parseRunOptions(['--model', 'opencode/gpt-5'], {
      [DEFAULT_MODEL_ENV]: 'opencode/env-default-free',
    });
    expect(argv.model).toBe('opencode/gpt-5');
  });

  test('json_standard_default', async () => {
    const argv = await parseRunOptions([]);
    expect(argv.jsonStandard).toBe('opencode');
  });

  test('json_standard_claude', async () => {
    const argv = await parseRunOptions(['--json-standard', 'claude']);
    expect(argv.jsonStandard).toBe('claude');
  });

  test('input_format_default', async () => {
    const argv = await parseRunOptions([]);
    expect(argv.inputFormat).toBe('text');
  });

  test('input_format_stream_json', async () => {
    const argv = await parseRunOptions(['--input-format', 'stream-json']);
    expect(argv.inputFormat).toBe('stream-json');
  });

  test('input_format_rejects_invalid', async () => {
    await expect(parseRunOptions(['--input-format', 'json'])).rejects.toThrow(
      'Invalid values'
    );
  });

  test('output_format_stream_json_maps_to_claude', async () => {
    const argv = await parseRunOptions(['--output-format', 'stream-json']);
    expect(argv.outputFormat).toBe('stream-json');
    expect(argv.jsonStandard).toBe('claude');
  });

  test('output_format_json_maps_to_opencode', async () => {
    const argv = await parseRunOptions([
      '--json-standard',
      'claude',
      '--output-format',
      'json',
    ]);
    expect(argv.outputFormat).toBe('json');
    expect(argv.jsonStandard).toBe('opencode');
  });

  test('json_standard_rejects_invalid', async () => {
    await expect(parseRunOptions(['--json-standard', 'xml'])).rejects.toThrow(
      'Invalid values'
    );
  });

  test('system_message_option', async () => {
    const argv = await parseRunOptions([
      '--system-message',
      'You are a test bot',
    ]);
    expect(argv.systemMessage).toBe('You are a test bot');
  });

  test('system_message_file_option', async () => {
    const argv = await parseRunOptions([
      '--system-message-file',
      '/tmp/sys.txt',
    ]);
    expect(argv.systemMessageFile).toBe('/tmp/sys.txt');
  });

  test('append_system_message_option', async () => {
    const argv = await parseRunOptions([
      '--append-system-message',
      'Extra instructions',
    ]);
    expect(argv.appendSystemMessage).toBe('Extra instructions');
  });

  test('append_system_message_file_option', async () => {
    const argv = await parseRunOptions([
      '--append-system-message-file',
      '/tmp/append.txt',
    ]);
    expect(argv.appendSystemMessageFile).toBe('/tmp/append.txt');
  });

  test('system_message_file_not_found', async () => {
    const argv = await parseRunOptions([
      '--system-message-file',
      '/tmp/nonexistent_file_12345.txt',
    ]);
    expect(argv.systemMessageFile).toBe('/tmp/nonexistent_file_12345.txt');
  });

  test('server_mode_default_true', async () => {
    const argv = await parseRunOptions([]);
    expect(argv.server).toBe(true);
  });

  test('server_mode_disabled', async () => {
    const argv = await parseRunOptions(['--no-server']);
    expect(argv.server).toBe(false);
  });

  test('verbose_shows_config', async () => {
    const argv = await parseRunOptions(['--verbose']);
    expect(argv.verbose).toBe(true);
    expect(argv.model).toBe('opencode/minimax-m2.5-free');
    expect(argv.jsonStandard).toBe('opencode');
    expect(argv.compactionModel).toBe('opencode/gpt-5-nano');
    expect(argv.compactionSafetyMargin).toBe(25);
  });

  test('verbose_off_hides_config', async () => {
    const argv = await parseRunOptions([]);
    expect(argv.verbose).toBe(false);
  });

  test('dry_run_echoes_message', async () => {
    const argv = await parseRunOptions(['--dry-run', '-p', 'test message']);
    expect(argv.dryRun).toBe(true);
    expect(argv.prompt).toBe('test message');
  });

  test('use_existing_claude_oauth_accepted', async () => {
    const argv = await parseRunOptions(['--use-existing-claude-oauth']);
    expect(argv.useExistingClaudeOauth).toBe(true);
  });

  test('prompt_short_flag', async () => {
    const argv = await parseRunOptions(['-p', 'short flag test']);
    expect(argv.prompt).toBe('short flag test');
  });

  test('prompt_long_flag', async () => {
    const argv = await parseRunOptions(['--prompt', 'long flag test']);
    expect(argv.prompt).toBe('long flag test');
  });

  test('disable_stdin_with_prompt_succeeds', async () => {
    const argv = await parseRunOptions(['--disable-stdin', '-p', 'hello']);
    expect(argv.disableStdin).toBe(true);
    expect(argv.prompt).toBe('hello');
  });

  test('disable_stdin_without_prompt_fails', async () => {
    const argv = await parseRunOptions(['--disable-stdin']);
    expect(argv.disableStdin).toBe(true);
    expect(argv.prompt).toBeUndefined();
  });

  test('stdin_stream_timeout_accepted', async () => {
    const argv = await parseRunOptions([
      '--stdin-stream-timeout',
      '5000',
      '-p',
      'hello',
    ]);
    expect(argv.stdinStreamTimeout).toBe(5000);
  });

  test('auto_merge_queued_messages_default', async () => {
    const argv = await parseRunOptions([]);
    expect(argv.autoMergeQueuedMessages).toBe(true);
  });

  test('no_auto_merge_queued_messages', async () => {
    const argv = await parseRunOptions(['--no-auto-merge-queued-messages']);
    expect(argv.autoMergeQueuedMessages).toBe(false);
  });

  test('interactive_default_true', async () => {
    const argv = await parseRunOptions([]);
    expect(argv.interactive).toBe(true);
  });

  test('no_interactive', async () => {
    const argv = await parseRunOptions(['--no-interactive']);
    expect(argv.interactive).toBe(false);
  });

  test('no_always_accept_stdin', async () => {
    const argv = await parseRunOptions(['--no-always-accept-stdin']);
    expect(argv.alwaysAcceptStdin).toBe(false);
  });

  test('compact_json_single_line', async () => {
    const argv = await parseRunOptions(['--compact-json']);
    expect(argv.compactJson).toBe(true);
  });

  test('resume_option_accepted', async () => {
    const argv = await parseRunOptions(['--resume', 'ses_abc123']);
    expect(argv.resume).toBe('ses_abc123');
  });

  test('resume_short_flag', async () => {
    const argv = await parseRunOptions(['-r', 'ses_abc123']);
    expect(argv.resume).toBe('ses_abc123');
  });

  test('continue_option_accepted', async () => {
    const argv = await parseRunOptions(['--continue']);
    expect(argv.continue).toBe(true);
  });

  test('continue_short_flag', async () => {
    const argv = await parseRunOptions(['-c']);
    expect(argv.continue).toBe(true);
  });

  test('no_fork_option_accepted', async () => {
    const argv = await parseRunOptions([
      '--no-fork',
      '--resume',
      'ses_abc',
      '-p',
      'hello',
    ]);
    expect(argv.noFork).toBe(true);
  });

  test('generate_title_option', async () => {
    const argv = await parseRunOptions(['--generate-title']);
    expect(argv.generateTitle).toBe(true);
  });

  test('retry_timeout_option', async () => {
    const argv = await parseRunOptions(['--retry-timeout', '3600']);
    expect(argv.retryTimeout).toBe(3600);
  });

  test('retry_on_rate_limits_default_true', async () => {
    const argv = await parseRunOptions([]);
    expect(argv.retryOnRateLimits).toBe(true);
  });

  test('no_retry_on_rate_limits', async () => {
    const argv = await parseRunOptions(['--no-retry-on-rate-limits']);
    expect(argv.retryOnRateLimits).toBe(false);
  });

  test('output_response_model_accepted', async () => {
    const argv = await parseRunOptions(['--no-output-response-model']);
    expect(argv.outputResponseModel).toBe(false);
  });

  test('summarize_session_default', async () => {
    const argv = await parseRunOptions([]);
    expect(argv.summarizeSession).toBe(true);
  });

  test('no_summarize_session', async () => {
    const argv = await parseRunOptions(['--no-summarize-session']);
    expect(argv.summarizeSession).toBe(false);
  });

  test('compaction_model_default', async () => {
    const argv = await parseRunOptions([]);
    expect(argv.compactionModel).toBe('opencode/gpt-5-nano');
  });

  test('compaction_model_custom', async () => {
    const argv = await parseRunOptions(['--compaction-model', 'same']);
    expect(argv.compactionModel).toBe('same');
  });

  test('compaction_model_env_default', async () => {
    const argv = await parseRunOptions([], {
      [DEFAULT_COMPACTION_MODEL_ENV]: 'opencode/env-compact-free',
    });
    expect(argv.compactionModel).toBe('opencode/env-compact-free');
  });

  test('compaction_model_cli_overrides_env_default', async () => {
    const argv = await parseRunOptions(['--compaction-model', 'same'], {
      [DEFAULT_COMPACTION_MODEL_ENV]: 'opencode/env-compact-free',
    });
    expect(argv.compactionModel).toBe('same');
  });

  test('compaction_models_default', async () => {
    const argv = await parseRunOptions([]);
    expect(argv.compactionModels).toBe(
      '(big-pickle minimax-m2.5-free nemotron-3-super-free hy3-preview-free ling-2.6-flash-free gpt-5-nano same)'
    );
  });

  test('compaction_models_custom', async () => {
    const argv = await parseRunOptions([
      '--compaction-models',
      '(model1 same)',
    ]);
    expect(argv.compactionModels).toBe('(model1 same)');
  });

  test('compaction_models_env_default', async () => {
    const argv = await parseRunOptions([], {
      [DEFAULT_COMPACTION_MODELS_ENV]: '(env-compact-free same)',
    });
    expect(argv.compactionModels).toBe('(env-compact-free same)');
  });

  test('compaction_models_cli_overrides_env_default', async () => {
    const argv = await parseRunOptions(
      ['--compaction-models', '(model1 same)'],
      { [DEFAULT_COMPACTION_MODELS_ENV]: '(env-compact-free same)' }
    );
    expect(argv.compactionModels).toBe('(model1 same)');
  });

  test('compaction_safety_margin_default', async () => {
    const argv = await parseRunOptions([]);
    expect(argv.compactionSafetyMargin).toBe(25);
  });

  test('compaction_safety_margin_custom', async () => {
    const argv = await parseRunOptions(['--compaction-safety-margin', '25']);
    expect(argv.compactionSafetyMargin).toBe(25);
  });

  test('compaction_safety_margin_env_default', async () => {
    const argv = await parseRunOptions([], {
      [DEFAULT_COMPACTION_SAFETY_MARGIN_PERCENT_ENV]: '12',
    });
    expect(argv.compactionSafetyMargin).toBe(12);
  });

  test('compaction_safety_margin_cli_overrides_env_default', async () => {
    const argv = await parseRunOptions(['--compaction-safety-margin', '25'], {
      [DEFAULT_COMPACTION_SAFETY_MARGIN_PERCENT_ENV]: '12',
    });
    expect(argv.compactionSafetyMargin).toBe(25);
  });

  test('all_options_accepted_together', async () => {
    const argv = await parseRunOptions([
      '--model',
      'opencode/gpt-5',
      '--json-standard',
      'claude',
      '--input-format',
      'stream-json',
      '--output-format',
      'stream-json',
      '--system-message',
      'Be helpful',
      '--verbose',
      '--dry-run',
      '--use-existing-claude-oauth',
      '--compact-json',
      '--no-interactive',
      '--no-always-accept-stdin',
      '--no-auto-merge-queued-messages',
      '--generate-title',
      '--no-retry-on-rate-limits',
      '--retry-timeout',
      '60',
      '--no-output-response-model',
      '--no-summarize-session',
      '--compaction-model',
      'same',
      '--compaction-models',
      '(same)',
      '--compaction-safety-margin',
      '20',
      '--temperature',
      '0.5',
      '--no-server',
      '--no-fork',
      '--resume',
      'ses_abc',
      '--stdin-stream-timeout',
      '1000',
      '--disable-stdin',
      '-p',
      'test',
    ]);

    expect(argv.model).toBe('opencode/gpt-5');
    expect(argv.jsonStandard).toBe('claude');
    expect(argv.inputFormat).toBe('stream-json');
    expect(argv.outputFormat).toBe('stream-json');
    expect(argv.systemMessage).toBe('Be helpful');
    expect(argv.verbose).toBe(true);
    expect(argv.dryRun).toBe(true);
    expect(argv.useExistingClaudeOauth).toBe(true);
    expect(argv.compactJson).toBe(true);
    expect(argv.interactive).toBe(false);
    expect(argv.alwaysAcceptStdin).toBe(false);
    expect(argv.autoMergeQueuedMessages).toBe(false);
    expect(argv.generateTitle).toBe(true);
    expect(argv.retryOnRateLimits).toBe(false);
    expect(argv.retryTimeout).toBe(60);
    expect(argv.outputResponseModel).toBe(false);
    expect(argv.summarizeSession).toBe(false);
    expect(argv.compactionModel).toBe('same');
    expect(argv.compactionModels).toBe('(same)');
    expect(argv.compactionSafetyMargin).toBe(20);
    expect(argv.temperature).toBe(0.5);
    expect(argv.server).toBe(false);
    expect(argv.noFork).toBe(true);
    expect(argv.resume).toBe('ses_abc');
    expect(argv.stdinStreamTimeout).toBe(1000);
    expect(argv.disableStdin).toBe(true);
    expect(argv.prompt).toBe('test');
  });
});
